src/rpc/blockchain.cpp

```diff
@@ -2126,6 +2128,152 @@ static UniValue getcolor(const JSONRPCRequest& request)
     return colorId.toHexString();
 }
 
+static void CreateUTXOSnapshot(
+    CAutoFile& afile,
+    const fs::path& path,
+    const fs::path& temppath,
+    UniValue& result)
+{
+    std::unique_ptr<CCoinsViewCursor> pcursor;
+    const CBlockIndex* tip;
+    CCoinsStats maybe_stats;
+
+    {
+        // We need to lock cs_main to ensure that the coinsdb isn't written to
+        // between (i) flushing coins cache to disk (coinsdb), (ii) getting stats
+        // based upon the coinsdb, and (iii) constructing a cursor to the
+        // coinsdb for use below this block.
+        //
+        // Cursors returned by leveldb iterate over snapshots, so the contents
+        // of the pcursor will not be affected by simultaneous writes during
+        // use below this block.
+        //
+        // See discussion here:
+        //   https://github.com/bitcoin/bitcoin/pull/15606#discussion_r274479369
+        //
+        LOCK(::cs_main);
+
+        FlushStateToDisk();
+
+        auto success = GetUTXOStats(pcoinsdbview.get(), maybe_stats);
+        if (!success) {
+            throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to read UTXO set");
+        }
+
+        pcursor = std::unique_ptr<CCoinsViewCursor>(pcoinsdbview.get()->Cursor());
+        tip = LookupBlockIndex(maybe_stats.hashBlock);
+    }
+
+    LogPrint(BCLog::RPC, "writing UTXO snapshot at height %d (%s) to file %s (via %s)\n",
+        tip->nHeight, tip->GetBlockHash().ToString().c_str(),
+        path, temppath);
+
+    SnapshotMetadata metadata{tip->GetBlockHash(), maybe_stats.nTransactionOutputs};
+
+    afile << metadata;
+
+    COutPoint key;
+    Coin coin;
+    unsigned int iter{0};
+    uint256 last_hash;
+    size_t written_coins_count{0};
+    std::vector<std::pair<uint32_t, Coin>> coins;
+
+    // To reduce space the serialization format of the snapshot avoids
+    // duplication of tx hashes. The code takes advantage of the guarantee by
+    // leveldb that keys are lexicographically sorted.
+    // In the coins vector we collect all coins that belong to a certain tx hash
+    // (key.hash) and when we have them all (key.hash != last_hash) we write
+    // them to file using the below lambda function.
+    // See also https://github.com/bitcoin/bitcoin/issues/25675
+    auto write_coins_to_file = [&](CAutoFile& afile, const uint256& last_hash, const std::vector<std::pair<uint32_t, Coin>>& coins, size_t& written_coins_count) {
+        afile << last_hash;
+        WriteCompactSize(afile, coins.size());
+        for (const auto& [n, coin] : coins) {
+            WriteCompactSize(afile, n);
+            afile << coin;
+            ++written_coins_count;
+        }
+    };
+    while (pcursor->Valid()) {
+        if (iter % 5000 == 0) ShutdownRequested();
```

Comment:
Why shut it down?

Reply:
This is the older version of checking for interrupts. In the bitcoin implementation there is a newer technique, `RpcInterruptionPoint`. In Tapyrus, we are checking whether to shutdown using `ShutdownRequested` at regular intervals.